contracts/OwnedSet.sol

```diff
@@ -12,35 +12,49 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-pragma solidity ^0.4.21;
+pragma solidity ^0.4.22;
 
-import "./interfaces/Owned.sol";
-import "./interfaces/ValidatorSet.sol";
+import "./Owned.sol";
+import "./ValidatorSet.sol";
 
 
 // Owner can add or remove validators.
 contract OwnedSet is Owned, ValidatorSet {
+	// TYPES
+	struct AddressStatus {
+		bool isIn;
+		uint index;
+	}
+
 	// EVENTS
 	event Report(address indexed reporter, address indexed reported, bool indexed malicious);
 	event ChangeFinalized(address[] currentSet);
 
+	// STATE
+
+	// System address, used by the block sealer.
+	address constant SYSTEM_ADDRESS = 0xffffFFFfFFffffffffffffffFfFFFfffFFFfFFfE;
+	uint public recentBlocks = 20;
+
+	// Current list of addresses entitled to participate in the consensus.
+	address[] validators;
+	address[] pending;
+	mapping(address => AddressStatus) pendingStatus;
+
+	// Was the last validator change finalized. Implies validators == pending
+	bool public finalized;
+
 	// MODIFIERS
 	modifier onlySystemAndNotFinalized() {
-		require(msg.sender != SYSTEM_ADDRESS || finalized);
+		require(msg.sender == SYSTEM_ADDRESS && !finalized);
 		_;
 	}
 
 	modifier whenFinalized() {
-		require(!finalized);
+		require(finalized);
 		_;
 	}
 
-	modifier isValidator(address _someone) {
-		if (pendingStatus[_someone].isIn) {
```

Comment:
I removed this since the only difference from `isPending` is that it doesn't throw. Since this is called from the engine do you know if we should make sure those calls don't throw?

Reply:
Even if it makes sense from code point of view, it sorta ruing the semantics of the calls (like on [line 149](https://github.com/parity-contracts/kovan-validator-set/pull/7/files#diff-7d0e17edb2208b50946ef7ca508ac60fR149)).

Generally it feels a little bit weird, why does a check `isValidator` checks for `pendingStatus`. Maybe let's rename that mapping as well, and then call the check `isValidator`?

Reply:
I agree with you. The fact that `isIn` is true doesn't guarantee the address is currently a validator because the change may have not been finalized yet. But when this is called from `addValidator` or `removeValidator` we also make sure that there aren't any pending changes (`whenFinalized`) so it's guaranteed that everything in `pendingStatus` has been finalized. For misbehavior reports it's necessary to do an extra check on the validators list so we could do something like `require(pendingStatus[_someone].isIn && validators[pendingStatus[_someone].index] == _someone);`. I'm not sure what to rename the mapping to, maybe just `status`? Since you actually also need to check `validators[]` to check whether it's pending or not.